include/zephyr/senss/senss_datatypes.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SENSS_DATATYPES_H_
+#define SENSS_DATATYPES_H_
+
+/**
+ * @brief Data Types
+ * @addtogroup senss_datatypes
+ * @{
+ */
+
+/**
+ * @struct senss_sensor_value_header
+ * @brief sensor value header
+ *
+ * Each sensor value data structure should have this header
+ *
+ * Here use 'base_timestamp' (uint64_t) and 'timestampe_delta' (uint32_t) to
+ * save memory usage in batching mode.
+ *
+ * The 'base_timestamp' is for readings[0], the 'timestamp_delta' is relation
+ * to the previous 'readings'. So,
+ * timestamp of readings[0] is
+ *	header.base_timestamp + readings[0].timestamp_delta.
+ * timestamp of readings[1] is
+ *	timestamp of readings[0] + readings[1].timestamp_delta.
+ *
+ * Since timestamp unit is nano seconds, the max 'timestamp_delta' (uint32_t)
+ * is 4.295 seconds.
+ *
+ * If a sensor has batched data where two consecutive readings differ by
+ * more than 4.295 seconds, the sensor subsystem core will split them
+ * across multiple instances of the readings structure, and send multiple
+ * events.
+ *
+ * This concept is borrowed from CHRE:
+ * https://cs.android.com/android/platform/superproject/+/master:\
+ * system/chre/chre_api/include/chre_api/chre/sensor_types.h
+ */
+struct senss_sensor_value_header {
+	/** base timestamp of this data readings, unit is nano seconds */
+	uint64_t base_timestamp;
+	/** count of this data readings */
+	uint16_t reading_count;
+};
+
+/**
+ * Sensor value data structure types based on common data types.
+ * Suitable for common sensors, such as IMU, Light sensors and orientation sensors.
+ *
+ * Sensor subsystem will use scaled fixed point data structure for all sensor values,
+ * aligned the HID spec, using the format v*10^x to present the decimal value,
+ * where the v is integer number, either int8/uint8, int16/uint6, or int32/uint32, depends on
+ * required sensor data precision.
+ *
+ * The scale unit exponent x is int8 type int8 type with encoding meanings (page 68 of
+ * HID spec https://usb.org/sites/default/files/hutrr39b_0.pdf):
```

Comment:
This is a pretty nice way of representing things, and surely makes doing math simpler though with the potential of an overflow/underflow scenario. Then again can always upscale to the next integer size or reduce the precision pretty easily I imagine if needed then.

Reply:
Yes!